### ISSUE TITLE
RootErrorFallback has incorrect "Error" interface for some errors and is therefore casted to "any"

### ISSUE THREAD

**Issue by egonm12** [blitz-js/blitz#1356]
### What is the problem?
This problem consists of two issues:
1. The type of `error` in the interface `FallbackProps` from `react-error-boundary` is inconsistent with the actual error the `RootErrorFallback` contains in some cases. Because of this, the errors that are handled inside the `RootErrorFallback` component are therefore casted to `any`. E.g.:
```
 return (
      <ErrorComponent
        statusCode={(error as any).statusCode}
        title="Sorry, you are not authorized to access this"
      />
    )
```

2. A fallback for an `AuthenticationError` is implemented inside the `RootErrorFallback` component but it will never get there because inside `LoginForm` itself it returns another error when an Authentication Error arises.

### Steps to Reproduce
1.  Let the ErrorBoundary component handle the error of login form:
In `LoginForm.tsx` add the following inside the component
```
import { useErrorHandler } from "react-error-boundary"

export const LoginForm = (props: LoginFormProps) => {
  const handleError = useErrorHandler();
  
  ...
}
```
2. Replace the code in the `catch` block of the `onSubmit` with:
```
catch (error) {
  handleError(error)
}
```
3. Go to `_app.tsx` and add a log inside the `RootErrorFallback` method.
```
console.log(JSON.stringify(error))
```
4. Go to the login page and enter an incorrect email address and submit it
5. See in the console that the `error` object has a different interface than defined by `FallbackProps` : 
```
{"name":"AuthenticationError","statusCode":401,"message":"You must be logged in to access this"}
```
As already explained, the `error` inside `RootErrorFallback` has a different interface defined.

### Versions

```
debug: blitzPkgPath: /Users/egonmeijers/projects/todo-app/node_modules/blitz/dist/index.js
debug: cliPkgPath: /Users/egonmeijers/projects/todo-app/node_modules/@blitzjs/cli/lib/src/index.js

macOS Catalina | darwin-x64 | Node: v12.16.1

blitz: 0.24.2 (global)
blitz: 0.24.2 (local)

  Package manager: yarn
  System:
    OS: macOS 10.15.7
    CPU: (12) x64 Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
    Memory: 226.69 MB / 16.00 GB
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 12.16.1 - ~/.nvm/versions/node/v12.16.1/bin/node
    Yarn: 1.22.4 - /usr/local/bin/yarn
    npm: 6.13.4 - ~/.nvm/versions/node/v12.16.1/bin/npm
    Watchman: Not Found
  npmPackages:
    @prisma/cli: 2.9.0 => 2.9.0
    @prisma/client: 2.9.0 => 2.9.0
    blitz: 0.24.2 => 0.24.2
    react: 0.0.0-experimental-7f28234f8 => 0.0.0-experimental-7f28234f8
    react-dom: 0.0.0-experimental-7f28234f8 => 0.0.0-experimental-7f28234f8
    typescript: 4.0.3 => 4.0.3
```

### Other

Please include applicable logs and screenshots that show your problem.

**Comment by flybayer** [comment]
Do you have a suggestion for how to improve this?

And fyi, the type of `error` inside `RootErrorFallback` comes from react-error-boundary.

**Comment by egonm12** [comment]
@flybayer I understand that the type of `error` comes from `react-error-boundary` so we have two options here:
- Option 1: Create a new type based on the type of `react-error-boundary` but extend it with the other possible error type(s). Probably something like
```
interface WithCustomErrorFallbackProps<CustomError> extends Omit<FallbackProps, 'error'> {
  error: FallbackProps['error'] | CustomError
}
```

With just this solution you are already able to cast to the `CustomErrorType` instead of `any` which already is much better.

- Option 2:

If we want to make this even more robuust we also need to update the interfaces for the custom blitz errors (authentication, authorization, etc). At the moment they all look the same (except for `CSRFTokenMismatchError`):
```
    name: string;
    statusCode: number;
    constructor(message?: string);
    get _clearStack(): boolean;
```

My suggestion would be to update the `name` properties for each error since its value is a fixed string and not 'any' string. E.g. 
```
export declare class AuthenticationError extends Error {
    name: 'AuthorizationError';
    statusCode: number;
    constructor(message?: string);
    get _clearStack(): boolean;
}
```
This way we can replace the string check inside the `RootErrorFallback` with a type guard like this:
```
interface WithCustomErrorFallbackProps<CustomError> extends Omit<FallbackProps, 'error'> {
  error: FallbackProps['error'] | CustomError
}

const isAuthenticationError = (maybeAuthenticationError: any): 
  maybeAuthenticationError is AuthenticationError => maybeAuthenticationError.name === 'AuthenticationError'
const isAuthorizationError = (maybeAuthorizationError: any): 
  maybeAuthorizationError is AuthorizationError => maybeAuthorizationError.name === 'AuthorizationError'

function RootErrorFallback({ error, resetErrorBoundary }: WithCustomErrorFallbackProps<AuthenticationError | AuthorizationError>) {
  if (isAuthenticationError(error)) {
    return <LoginForm onSuccess={resetErrorBoundary} />
  } else if (isAuthorizationError(error)) {
    return (
      <ErrorComponent
        statusCode={error.statusCode} // <-- this is now scoped to a AuthenticationError type
        title="Sorry, you are not authorized to access this"
      />
    )
  } else {
   ...
  }
}
```

**Comment by egonm12** [comment]
@flybayer I am willing to create a PR for this improvement.

**Comment by flybayer** [comment]
@egonm12 ok yeah, good ideas!

Currently the type guard functions would be needed, but I think once we update superjson in blitz to be the latest and use superjson for the errors, we can use `instanceof` to check against the built-in blitz errors.

**Comment by egonm12** [comment]
@flybayer great suggestion. I think it is already possible to use the `instanceof` type guard. Why would updating superjson be needed according to you?

**Comment by flybayer** [comment]
Closed in #1414
